lib/nostr.js

```diff
@@ -200,4 +217,65 @@ class NDKNip46SignerURLPatch extends NDKNip46Signer {
     connectionToken = connectionToken.replace('bunker://', 'http://')
     return super.connectionTokenInit(connectionToken)
   }
+
+  async blockUntilReady () {
+    if (this.nip05 && !this.userPubkey) {
+      const user = await NDKUser.fromNip05(this.nip05, this.ndk)
+      if (user) {
+        this._user = user
+        this.userPubkey = user.pubkey
+        this.relayUrls = user.nip46Urls
+        this.rpc = new NDKNostrRpc(this.ndk, this.localSigner, this.debug, this.relayUrls)
+      }
+    }
+    if (!this.bunkerPubkey && this.userPubkey) {
+      this.bunkerPubkey = this.userPubkey
+    } else if (!this.bunkerPubkey) {
+      throw new Error('Bunker pubkey not set')
+    }
+    await this.startListening()
+    this.rpc.on('authUrl', (...props) => {
+      this.emit('authUrl', ...props)
+    })
+    return new Promise((resolve, reject) => {
+      const connectParams = [this.userPubkey ?? '']
+      if (this.secret) connectParams.push(this.secret)
+      if (!this.bunkerPubkey) throw new Error('Bunker pubkey not set')
+
+      const confirm = async (resolve) => {
+        return withTimeout(this.getPublicKey().then((pubkey) => {
+          this._user = new NDKUser({ pubkey })
+          resolve(this._user)
+        }), 15_000)
+      }
+
+      // initiate a new connection
+      const connector = setTimeout(() => {
+        this.rpc.sendRequest(
+          this.bunkerPubkey,
+          'connect',
+          connectParams,
+          24133,
+          async (response) => {
+            try {
+              if (response.result === 'ack') {
+                await confirm(resolve)
+              } else {
+                throw new Error(response.error)
+              }
+            } catch (e) {
+              reject(e)
+            }
+          }
+        )
+      }, 2100)
+
+      // try to restore the connection
+      // if it succeeds, clear the timeout.
+      //    One of the two will fail, either the new connection or the restore attempt.
+      confirm(resolve)
+        .then(() => clearTimeout(connector))
+        .catch((err) => console.error('Error restoring connection:', err))
+    })
```

Comment:
The [spec](https://github.com/nostr-protocol/nips/blob/master/46.md) is not very clear about how to re-establish connections between user sessions. It only mentions that the client key pair is disposable and suggests storing it locally for the user session. It also states that new connections cannot be established with a reused secret.

The key point here is that nsec.app will respond to connections from an authorized client key pair but will reject new connect RPCs from the same secret and key pair. The issue is that the NDK auth flow doesn’t seem to account for this behavior.

To work around this, I implemented a solution where two requests,one for "connect" and one for "get_public_key", are raced against each other. The first request to complete without an error is used to proceed with authentication.